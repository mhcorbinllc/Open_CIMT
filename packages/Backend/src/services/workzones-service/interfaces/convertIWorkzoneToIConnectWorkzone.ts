import {IWorkzone} from "./IWorkzone";
import {IConnectWorkzone} from "./IConnectWorkzone";
import {IWorkzonesAppSettings} from "./IWorkzonesAppSettings";

export const convertIWorkzoneToIConnectWorkzone = (
  {
    workzone,
    workzoneAppSettings,
  }: {
    workzone: IWorkzone;
    workzoneAppSettings: IWorkzonesAppSettings;
  },
): IConnectWorkzone => {
  return {
    id: workzone.kapschId,
    name: workzone.name,
    speedlimit: Math.round(workzone.speedLimitInMiles),
    isWorkzone: true,
    areWorkersPresent: workzone.workersPresent,
    shoulder: workzone.closedShoulder,
    lane: workzone.closedLane,
    laneWidth: Math.round(workzone.closedLaneWidthInMeters),
    start: dateToIsoString(workzone.start),
    end: dateToIsoString(workzone.end),
    forEver: workzone.forEver,
    itisCodes: workzone.itisCodes,

    // Point
    point: {
      type: "Point",
      coordinates: [
        workzone.point.lng,
        workzone.point.lat,
        workzone.point.alt === undefined
          ? null
          : workzone.point.alt,
      ],
    },

    heading: workzone.heading,

    // Path
    usePath: !!workzone.path.points.length,
    path: {
      type: "LineString",
      coordinates:
        workzone.path.points
          .map(point => ([
            point.lng,
            point.lat,
            point.alt || 0,
          ])),
    },

    // Selection mode: Region
    broadcastRegion: {
      type: "Polygon",
      coordinates: [
        workzone.broadcast.selectionByRegion.region.points
          .map(point => ([
            point.lng,
            point.lat,
            point.alt || 0,
          ])),
      ],
    },

    // Selection mode: Radius
    useRadius: !!workzone.broadcast.selectionByRadius.radiusInMeters,
    radius: workzone.broadcast.selectionByRadius.radiusInMeters,

    // CIMs features options
    serviceChannelStr: workzoneAppSettings.serviceChannel,
    psid: workzoneAppSettings.psid,
    txmode: workzoneAppSettings.txmode,
    interval: workzoneAppSettings.intervalInMs,
    priority: workzoneAppSettings.priority,
    useSignature: workzoneAppSettings.useSignature,
    useEncryption: workzoneAppSettings.useEncryption,
  };
};

const dateToIsoString = (date: Date | string | number): string => {
  return new Date(date).toISOString()
    .split('.')[0] + "Z";
};
