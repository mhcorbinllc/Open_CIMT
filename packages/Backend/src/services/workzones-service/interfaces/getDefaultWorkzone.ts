import * as moment from "moment";

import {
  EWZBroadcastSelectionMode,
  EWZClosureType,
  EWZLane,
  EWZShoulder,
  IWorkzone,
} from "./IWorkzone";
import {
  EWZClosureSelectionMode,
} from "./IWorkzonesAppSettings";

export const getDefaultWorkzone = (): IWorkzone => {
  const now = moment().seconds(0);

  return {
    workzoneId: '',
    kapschId: -1,
    name: '',
    active: true,

    notes: '',
    start: now.toDate(),
    end: now.clone().add(3, 'days')
      .toDate(),
    forEver: false,

    closureSelectionMode: EWZClosureSelectionMode.SIMPLE,
    closureType: EWZClosureType.NONE,
    itisCodes: [],
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 0,

    closedLaneWidthInMeters: 0,
    workersPresent: false,

    point: {
      lat: 0,
      lng: 0,
      alt: 0,
    },

    path: {points: []},

    heading: [],

    broadcast: {
      selectionMode: EWZBroadcastSelectionMode.REGION,
      selectionByRegion: {region: {points: []}},
      selectionByRadius: {radiusInMeters: 0},
    },

    createdAt: 0,
    updatedAt: 0,
    deletedAt: 0,
  };
};

