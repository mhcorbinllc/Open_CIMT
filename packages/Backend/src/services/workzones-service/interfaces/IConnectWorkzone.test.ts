import {
  IWorkzone,
  EWZLane,
  EWZShoulder,
  EWZBroadcastSelectionMode,
} from "./IWorkzone";
import {
  IWorkzonesAppSettings,
  EWZTXMode,
  EWZClosureSelectionMode,
} from "./IWorkzonesAppSettings";

import {convertIWorkzoneToIConnectWorkzone} from "./convertIWorkzoneToIConnectWorkzone";
import {getDefaultWorkzone} from "./getDefaultWorkzone";

const workzoneAppSettings: IWorkzonesAppSettings = {
  closureSelectionMode: EWZClosureSelectionMode.SIMPLE,
  deviceId: "*",
  intervalInMs: 1300,
  itisCodes: [
    {
      code: 769,
      label: "Closed to traffic",
    },
    {
      code: 771,
      label: "Closed ahead",
    },
    {
      code: 7451,
      label: "Merge",
    },
    {
      code: 8195,
      label: "Left lane",
    },
    {
      code: 8196,
      label: "Right lane",
    },
    {
      code: 8208,
      label: "Right shoulder",
    },
    {
      code: 8209,
      label: "Left shoulder",
    },
    {
      code: 12545,
      label: "One",
    },
    {
      code: 12546,
      label: "Two",
    },
    {
      code: 12547,
      label: "Three",
    },
    {
      code: 12548,
      label: "Four",
    },
    {
      code: 12549,
      label: "Five",
    },
    {
      code: 13579,
      label: "Right",
    },
    {
      code: 13580,
      label: "Left",
    },
  ],
  priority: 62,
  psid: "8002F",
  serviceChannel: "176",
  txmode: EWZTXMode.CONT,
  useEncryption: true,
  useSignature: false,
  createdAt: 1614763319202,
  updatedAt: 1620226220921,
  deletedAt: 0,
};

const now = new Date(0).valueOf();

const validBasicWorkzone: IWorkzone = {
  ...getDefaultWorkzone(),
  active: true,
  start: new Date(now + 200),
  end: new Date(now + 201),
  name: 'Demo CIM',
  workzoneId: 'ed-98467204',
  itisCodes: [200, 300],
  closedLane: EWZLane.CENTER,
  closedShoulder: EWZShoulder.LEFT,
  closedLaneWidthInMeters: 12,
  point: {
    lng: 0.224423,
    lat: 0.134211,
  },

  path: {
    points: [
      {
        lng: 0.224423,
        lat: 0.134211,
      },
      {
        lng: 0.224423,
        lat: 0.134212,
      },
    ],
  },
  broadcast: {
    selectionMode: EWZBroadcastSelectionMode.RADIUS,
    selectionByRadius: {radiusInMeters: 12},
    selectionByRegion: {region: {points: []}},
  },
};

describe('convertIWorkzoneToIKapschWorkzone', () => {
  test('Convert to IKapschWorkzone a Radius CIM', () => {
    expect(
      convertIWorkzoneToIConnectWorkzone({
        workzoneAppSettings,
        workzone: JSON.parse(JSON.stringify(validBasicWorkzone as IWorkzone)),
      }),
    ).toMatchSnapshot();
  });
  test('Convert to IKapschWorkzone a Region CIM', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify(validBasicWorkzone as IWorkzone));
    workzone.kapschId = 32441;
    workzone.broadcast.selectionMode = EWZBroadcastSelectionMode.REGION;
    workzone.broadcast.selectionByRegion.region.points = [
      {
        lat: 0.23,
        lng: 0.45,
      },
      {
        lat: 0.23,
        lng: 0.46,
      },
      {
        lat: 0.24,
        lng: 0.46,
      },
      {
        lat: 0.23,
        lng: 0.47,
      },
    ];

    expect(
      convertIWorkzoneToIConnectWorkzone({
        workzoneAppSettings,
        workzone,
      }),
    ).toMatchSnapshot();
  });
});
