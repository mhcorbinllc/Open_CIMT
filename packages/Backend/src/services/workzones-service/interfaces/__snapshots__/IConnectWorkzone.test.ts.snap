// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertIWorkzoneToIKapschWorkzone Convert to IKapschWorkzone a Radius CIM 1`] = `
Object {
  "areWorkersPresent": false,
  "broadcastRegion": Object {
    "coordinates": Array [
      Array [],
    ],
    "type": "Polygon",
  },
  "end": "1970-01-01T00:00:00Z",
  "forEver": false,
  "heading": Array [],
  "id": -1,
  "interval": 1300,
  "isWorkzone": true,
  "itisCodes": Array [
    200,
    300,
  ],
  "lane": "CENTER",
  "laneWidth": 12,
  "name": "Demo CIM",
  "path": Object {
    "coordinates": Array [
      Array [
        0.224423,
        0.134211,
        0,
      ],
      Array [
        0.224423,
        0.134212,
        0,
      ],
    ],
    "type": "LineString",
  },
  "point": Object {
    "coordinates": Array [
      0.224423,
      0.134211,
      null,
    ],
    "type": "Point",
  },
  "priority": 62,
  "psid": "8002F",
  "radius": 12,
  "serviceChannelStr": "176",
  "shoulder": "LEFT",
  "speedlimit": 0,
  "start": "1970-01-01T00:00:00Z",
  "txmode": "CONT",
  "useEncryption": true,
  "usePath": true,
  "useRadius": true,
  "useSignature": false,
}
`;

exports[`convertIWorkzoneToIKapschWorkzone Convert to IKapschWorkzone a Region CIM 1`] = `
Object {
  "areWorkersPresent": false,
  "broadcastRegion": Object {
    "coordinates": Array [
      Array [
        Array [
          0.45,
          0.23,
          0,
        ],
        Array [
          0.46,
          0.23,
          0,
        ],
        Array [
          0.46,
          0.24,
          0,
        ],
        Array [
          0.47,
          0.23,
          0,
        ],
      ],
    ],
    "type": "Polygon",
  },
  "end": "1970-01-01T00:00:00Z",
  "forEver": false,
  "heading": Array [],
  "id": 32441,
  "interval": 1300,
  "isWorkzone": true,
  "itisCodes": Array [
    200,
    300,
  ],
  "lane": "CENTER",
  "laneWidth": 12,
  "name": "Demo CIM",
  "path": Object {
    "coordinates": Array [
      Array [
        0.224423,
        0.134211,
        0,
      ],
      Array [
        0.224423,
        0.134212,
        0,
      ],
    ],
    "type": "LineString",
  },
  "point": Object {
    "coordinates": Array [
      0.224423,
      0.134211,
      null,
    ],
    "type": "Point",
  },
  "priority": 62,
  "psid": "8002F",
  "radius": 12,
  "serviceChannelStr": "176",
  "shoulder": "LEFT",
  "speedlimit": 0,
  "start": "1970-01-01T00:00:00Z",
  "txmode": "CONT",
  "useEncryption": true,
  "usePath": true,
  "useRadius": true,
  "useSignature": false,
}
`;
